File: examples/php/ledger.php
<?php
/*
 * Copyright 2026 Chengxi Luo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * LedgerEntry mirrors examples/cases/ledger.yaml.
 *
 * The attributes are the entire schema binding — nothing here calls a get/set
 * accessor. Everything else is the byte layout, which is the part a conformance
 * worker exists to exercise.
 *
 * PHP's int is only 64-bit, so every 64/128-bit value is carried as a decimal
 * string and converted through the GMP helpers in wire.php.
 *
 * Go is the --ref language and owns the layout; see examples/go/wire.go.
 */

declare(strict_types=1);

use Serify\Attributes\SerifyField;
use Serify\Attributes\SerifyModel;

#[SerifyModel]
class LedgerEntry
{
    #[SerifyField] public string $entryId = '0';
    #[SerifyField] public string $blockNumber = '0';
    #[SerifyField] public string $blockTime = '0';
    #[SerifyField] public string $txHash = '';
    #[SerifyField] public string $account = '';
    #[SerifyField] public string $asset = '';
    #[SerifyField] public string $amountBaseUnits = '0';
    #[SerifyField] public string $balanceAfter = '0';
    #[SerifyField] public bool $confirmed = false;
    #[SerifyField] public ?string $memo = null;

    public function marshal(): string
    {
        $out = encodeInt($this->entryId, 8)
             . encodeInt($this->blockNumber, 8)
             . encodeInt($this->blockTime, 8);

        $out .= lenPrefixed($this->txHash);
        $out .= lenPrefixed($this->account);
        $out .= lenPrefixed($this->asset);

        $out .= encodeInt($this->amountBaseUnits, 16);
        $out .= encodeInt($this->balanceAfter, 16);
        $out .= pack('C', $this->confirmed ? 1 : 0);

        if ($this->memo === null) {
            $out .= pack('C', 0);
        } else {
            $out .= pack('C', 1) . lenPrefixed($this->memo);
        }

        return $out;
    }

    public static function unmarshal(string $data): self
    {
        $e = new self();

        $e->entryId     = decodeUnsigned(substr($data, 0, 8));
        $e->blockNumber = decodeUnsigned(substr($data, 8, 8));
        $e->blockTime   = decodeSigned(substr($data, 16, 8));
        $off = 24;

        $e->txHash  = readLenPrefixed($data, $off);
        $e->account = readLenPrefixed($data, $off);
        $e->asset   = readLenPrefixed($data, $off);

        $e->amountBaseUnits = decodeSigned(substr($data, $off, 16));
        $e->balanceAfter    = decodeSigned(substr($data, $off + 16, 16));
        $off += 32;

        $e->confirmed = ord($data[$off]) !== 0;
        $hasMemo = ord($data[$off + 1]) !== 0;
        $off += 2;
        $e->memo = $hasMemo ? readLenPrefixed($data, $off) : null;

        return $e;
    }
}

File: examples/csharp/Ledger.cs
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// LedgerEntry mirrors examples/cases/ledger.yaml.
//
// [SerifyModel] plus one [SerifyField] per property is the entire schema binding
// — nothing here calls a Get*/Set* accessor. Everything else is the byte layout,
// which is the part a conformance worker exists to exercise.
//
// .NET has native Int128 and BinaryPrimitives can write it directly, so the two
// int128 fields need no big-integer library — unlike Go (math/big) or Java
// (BigInteger).
//
// Go is the --ref language and owns the layout; see examples/go/wire.go.

using System;
using System.Buffers.Binary;
using System.IO;
using Serify;

[SerifyModel]
internal sealed class LedgerEntry
{
    [SerifyField] public ulong EntryId { get; set; }
    [SerifyField] public ulong BlockNumber { get; set; }
    [SerifyField] public long BlockTime { get; set; }
    [SerifyField] public byte[] TxHash { get; set; } = Array.Empty<byte>();
    [SerifyField] public string Account { get; set; } = "";
    [SerifyField] public string Asset { get; set; } = "";
    [SerifyField] public Int128 AmountBaseUnits { get; set; }
    [SerifyField] public Int128 BalanceAfter { get; set; }
    [SerifyField] public bool Confirmed { get; set; }
    [SerifyField] public string? Memo { get; set; }

    public byte[] Marshal()
    {
        var ms = new MemoryStream();
        Span<byte> buf = stackalloc byte[16];

        BinaryPrimitives.WriteUInt64LittleEndian(buf, EntryId);
        ms.Write(buf[..8]);
        BinaryPrimitives.WriteUInt64LittleEndian(buf, BlockNumber);
        ms.Write(buf[..8]);
        BinaryPrimitives.WriteInt64LittleEndian(buf, BlockTime);
        ms.Write(buf[..8]);

        Wire.WriteLenPrefixed(ms, TxHash);
        Wire.WriteLenPrefixed(ms, Account);
        Wire.WriteLenPrefixed(ms, Asset);

        BinaryPrimitives.WriteInt128LittleEndian(buf, AmountBaseUnits);
        ms.Write(buf);
        BinaryPrimitives.WriteInt128LittleEndian(buf, BalanceAfter);
        ms.Write(buf);

        ms.WriteByte(Confirmed ? (byte)1 : (byte)0);

        if (Memo is null)
        {
            ms.WriteByte(0);
        }
        else
        {
            ms.WriteByte(1);
            Wire.WriteLenPrefixed(ms, Memo);
        }

        return ms.ToArray();
    }

    public static LedgerEntry Unmarshal(byte[] data)
    {
        ReadOnlySpan<byte> span = data;
        var e = new LedgerEntry
        {
            EntryId     = BinaryPrimitives.ReadUInt64LittleEndian(span),
            BlockNumber = BinaryPrimitives.ReadUInt64LittleEndian(span[8..]),
            BlockTime   = BinaryPrimitives.ReadInt64LittleEndian(span[16..]),
        };
        int off = 24;

        // ToArray copies: a slice would alias the input buffer.
        e.TxHash  = Wire.ReadLenPrefixed(span, ref off).ToArray();
        e.Account = Wire.ReadLenString(span, ref off);
        e.Asset   = Wire.ReadLenString(span, ref off);

        e.AmountBaseUnits = BinaryPrimitives.ReadInt128LittleEndian(span[off..]);
        e.BalanceAfter    = BinaryPrimitives.ReadInt128LittleEndian(span[(off + 16)..]);
        off += 32;

        e.Confirmed = span[off] != 0;
        bool hasMemo = span[off + 1] != 0;
        off += 2;
        e.Memo = hasMemo ? Wire.ReadLenString(span, ref off) : null;

        return e;
    }
}
